packages/eslint-plugin/src/rules/no-state-access.ts

```diff
@@ -0,0 +1,112 @@
+import * as ts from "@typescript-eslint/eslint-plugin";
+import * as tsPluginUtil from "@typescript-eslint/eslint-plugin/dist/util";
+import {
+  ParserServices,
+  TSESTree,
+} from "@typescript-eslint/experimental-utils";
+import { AST_NODE_TYPES, TSNode } from "@typescript-eslint/typescript-estree";
+import { TSRuleContext, TSRuleModule } from "../types/rules";
+import { matchModel } from "../utils/matchModel";
+
+const rule: TSRuleModule = {
+  meta: {
+    docs: {
+      description:
+        "Requires that the state variable is only accessed inside a component or action",
+      category: "Possible Errors",
+      recommended: true,
+    },
+    messages: {
+      accessingState:
+        "Trying to access the state variable. This is empty outside components or functions.",
+    },
+    schema: [],
+    type: "problem",
+  },
+  create(context: TSRuleContext) {
+    const parserServices: ParserServices = tsPluginUtil.getParserServices(
+      context,
+    );
+    const checker = parserServices.program!.getTypeChecker();
```

Comment:
Can `program` ever be undefined? If not, why does the type think it can?

Reply:
From here: https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/typescript-estree/src/parser.ts. If not AST and program, it throws an error. Can still check for `undefined` though.